Assets/BossRoom/Scripts/Client/UI/UIQuitPanel.cs

```diff
@@ -1,41 +1,48 @@
+using System;
 using Unity.Multiplayer.Samples.BossRoom.Shared;
 using Unity.Multiplayer.Samples.BossRoom.Shared.Infrastructure;
-using Unity.Netcode;
 using UnityEngine;
-using UnityEngine.UI;
 
 namespace Unity.Multiplayer.Samples.BossRoom.Client
 {
     public class UIQuitPanel : MonoBehaviour
     {
+        enum QuitMode
+        {
+            ReturnToMenu,
+            QuitApplication
+        }
+
         [SerializeField]
-        Text m_QuitButtonText;
+        QuitMode m_QuitMode = QuitMode.ReturnToMenu;
 
         ApplicationController m_ApplicationController;
 
-        private bool m_QuitMode = true;
 
         [Inject]
-        private void InjectDependencies(ApplicationController applicationController)
+        void InjectDependencies(ApplicationController applicationController)
         {
             m_ApplicationController = applicationController;
         }
 
-        void OnEnable()
+        void Awake()
         {
-            m_QuitMode = NetworkManager.Singleton == null || !NetworkManager.Singleton.IsListening;
-            m_QuitButtonText.text = m_QuitMode ? "Exit Game?" : "Leave session?";
+            var scope = DIScope.RootScope;
+            scope.InjectIn(gameObject);
```

Comment:
I would much rather see this injection in ApplicationController - this way we are spreading initialization around. How do you know for sure that DIScope.RootScope is fully constructed at the point of Awake? Can't have that guarantee unless you have it in one spot.

Reply:
Makes sense. I'm not sure how to go about doing it though, since this is now moved to different scenes. It used to be done through ApplicationController's `m_GameObjectsThatWillBeInjectedAutomatically`. What would be the best way to do this without `m_GameObjectsThatWillBeInjectedAutomatically`?

Reply:
If we know for a fact that DIScope.RootScope is loaded and set up in a different scene - then we're good I guess.

Reply:
In that case, yes RootScope is set up before that because it is done in ApplicationController's Awake, and UIQuitPanel's Awake is called when its game object gets activated (which is when we first click on the Quit button, or when loading the MainMenu scene if we don't deactivate the Panel in the prefab. However, if we set the panel active in the prefab, and add the prefab to the Startup scene, there may be a possibility that this Awake is called before ApplicationController's. Should I move this to Start just to be sure?

Reply:
Start is safer in this case, sure!

Reply:
My thought process regarding overall maintainability of the codebase is that as much as possible should be explicit and in code (it's version-controlled, it's readable by humans). `Awake` and `Start` functions are called for us by Unity (thus allowing us to have arbitrary entry points to our logic), but we do have both ApplicationController and various state controllers - these are the "designated entry points". Designated entry point versus non-designated entry point boils down to difference in cognitive complexity - if everything in the app follows the convention of being set up in app/state controllers, then we can rely on that fact. Breaking this assumption would bring us back to a point where we have multiple places where logic starts. << that's me being a bit of a zealot

I understand that Start would work in this case (and it probably would even be safe), however you're relying on a side-effect that for some reason we chose to do initialization in Awake, not Start. To be specific -I don't think that this assumption will be broken, I think that your approach would work.

Reply:
Updated this PR based on the discussion we had. Updated the PR description as well to reflect this change